src/envoy/tcp/alpn_proxy/alpn_proxy.cc

```diff
@@ -0,0 +1,260 @@
+      tryReadProxyData(data);
+      if (conn_state_ == NeedMoreDataProxyHeader) {
+        return Network::FilterStatus::StopIteration;
+      }
+      if (conn_state_ == Invalid) {
+        return Network::FilterStatus::Continue;
+      }
+    }
+    default:
+      conn_state_ = Done;
+      return Network::FilterStatus::Continue;
+  }
+
+  return Network::FilterStatus::Continue;
+}
+
+Network::FilterStatus AlpnProxyFilter::onNewConnection() {
+  return Network::FilterStatus::Continue;
+}
+
+Network::FilterStatus AlpnProxyFilter::onWrite(Buffer::Instance&, bool) {
+  switch (conn_state_) {
+    case Invalid:
+    case Done:
+      // No work needed if connection state is Done or Invalid.
+      return Network::FilterStatus::Continue;
+    case ConnProtocolNotRead: {
+      if (read_callbacks_->connection().nextProtocol() != config_->protocol_) {
+        conn_state_ = Invalid;
+        config_->stats().alpn_protocol_not_found_.inc();
+        return Network::FilterStatus::Continue;
+      } else {
+        conn_state_ = WriteMetadata;
+        config_->stats().alpn_protocol_found_.inc();
+      }
+    }
+    case WriteMetadata: {
+      // TODO(gargnupur): Try to move this just after alpn protocol is
+      // determined and first onWrite is called in Upstream filter.
+      if (config_->filter_direction_ == FilterDirection::Upstream) {
+        writeNodeMetadata();
+      }
+    }
+    case ReadingInitialHeader:
+    case ReadingProxyHeader:
+    case NeedMoreDataInitialHeader:
+    case NeedMoreDataProxyHeader:
+      // These are to be handled in Reading Pipeline.
+      return Network::FilterStatus::Continue;
+  }
+
+  return Network::FilterStatus::Continue;
+}
+
+void AlpnProxyFilter::writeNodeMetadata() {
+  if (conn_state_ != WriteMetadata) {
+    return;
+  }
+
+  std::unique_ptr<const google::protobuf::Struct> metadata =
+      readMetadata(config_->node_metadata_id_);
+  if (metadata != nullptr) {
+    Envoy::ProtobufWkt::Any metadata_any_value;
+    *metadata_any_value.mutable_type_url() = StructTypeUrl;
+    *metadata_any_value.mutable_value() = metadata->SerializeAsString();
+    std::unique_ptr<::Envoy::Buffer::OwnedImpl> buf =
+        constructProxyHeaderData(metadata_any_value);
+    write_callbacks_->injectWriteDataToFilterChain(*buf, false);
+
+    if (config_->filter_direction_ == FilterDirection::Downstream) {
+      writeMetadata(DownstreamDynamicDataKey, *metadata);
+    } else {
+      writeMetadata(UpstreamDynamicDataKey, *metadata);
+    }
+    config_->stats().metadata_added_.inc();
+  }
+
+  conn_state_ = ReadingInitialHeader;
+}
+
+void AlpnProxyFilter::tryReadInitialProxyHeader(Buffer::Instance& data) {
+  if (conn_state_ != ReadingInitialHeader &&
+      conn_state_ != NeedMoreDataInitialHeader) {
+    return;
+  }
+  const uint32_t initial_header_length = sizeof(AlpnProxyInitialHeader);
+  if (data.length() < initial_header_length) {
+    config_->stats().initial_header_not_found_.inc();
```

Comment:
This does not seem right to increment header on buffering. This is mostly random depending how TCP data is chunked, right?

Reply:
Please check replied above...
I thought "return Network::FilterStatus::StopIteration;" is doing that and we don't need to drain "data" and buffer bytes read till now..